adapter/internal/sourcewatcher/sourcewatcher.go

```diff
@@ -0,0 +1,286 @@
+		loggers.LoggerAPI.Errorf("Error reading configs: %v", err)
+	}
+
+	retryInterval := conf.Adapter.SourceControl.RetryInterval
+
+	loggers.LoggerAPI.Info("Starting source watcher")
+	// Fetch the API artifacts at the startup
+	repository, err := fetchArtifacts()
+
+	// Retry fetching the API artifacts if the first attempt fails
+	for {
+		if err == nil {
+			break
+		} else {
+			if err != nil{
+				loggers.LoggerAPI.Errorf("Error while fetching API artifacts during startup: %v", err)
+			}
+			time.Sleep(time.Duration(retryInterval) * time.Second)
+			loggers.LoggerAPI.Info("Retrying fetching artifacts from the remote repository")
+			repository, err = fetchArtifacts()
+		}
+	}
+
+	artifactsMap, err = api.ProcessMountedAPIProjects()
+	if err != nil {
+		logger.LoggerMgw.Error("Readiness probe is not set as local api artifacts processing has failed.")
+	}
+
+	loggers.LoggerAPI.Info("Polling for changes")
+	go pollChanges(repository)
+}
+
+// fetchArtifacts clones the API artifacts from the remote repository into the artifacts directory in the adapter
+func fetchArtifacts() (repository *git.Repository,err error) {
+	conf, err := config.ReadConfigs()
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error reading configs: %v", err)
+		return nil, err
+	}
+
+	artifactsDirName := filepath.FromSlash(conf.Adapter.SourceControl.ArtifactsDirectory + "/" + apisArtifactDir)
+	repositoryURL := conf.Adapter.SourceControl.Repository.URL
+	branch := conf.Adapter.SourceControl.Repository.Branch
+
+	// Opens the local repository, if exists
+	repository, _ = git.PlainOpen(artifactsDirName)
+
+	// If a local repository exists, pull the changes from the remote repository
+	if repository != nil {
+		loggers.LoggerAPI.Info("Starting to fetch changes from remote repository")
+
+		compareRepository(repository)
+
+		return repository, nil
+	}
+
+	// If a local repository does not exist, clone the remote repository
+	gitAuth, err := auth.GetGitAuth()
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error while authenticating the remote repository: %v", err)
+		return nil, err
+	}
+
+	cloneOptions := &git.CloneOptions{
+		URL: repositoryURL,
+		Auth: gitAuth,
+	}
+
+	if branch != "" {
+		cloneOptions.ReferenceName = plumbing.ReferenceName("refs/heads/" + branch)
+	}
+
+	loggers.LoggerAPI.Info("Fetching API artifacts from remote repository")
+
+	// Clones the  remote repository
+	repository, err = git.PlainClone(artifactsDirName, false, cloneOptions)
+
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error while fetching artifacts from the remote repository: %v", err)
+	}
+
+	return repository, err
+
+}
+
+// pollChanges polls for changes from the remote repository
+func pollChanges(repository *git.Repository){
+	conf, err := config.ReadConfigs()
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error reading configs: %v", err)
+		return
+	}
+
+	pollInterval := conf.Adapter.SourceControl.PollInterval
+	for {
+		<- time.After(time.Duration(pollInterval) * time.Second)
```

Comment:
Shall we add a debug log here stating that it is trying to pull changes from the repo?

Reply:
Fixed with cdeb393